<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
   "http://www.w3.org/TR/html4/strict.dtd">

<html lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
	<title>untitled</title>
	<meta name="generator" content="TextMate http://macromates.com/">
	<meta name="author" content="macleand">
	<!-- Date: 2011-10-14 -->
	<link rel="stylesheet" type="text/css" href="basic_styles.css"></link>
	<link rel="stylesheet" type="text/css" href="http://drnicwilliams.com/external/CodeHighlighter/styles.css"></link>  
	<script src="http://drnicwilliams.com/external/CodeHighlighter/clean_tumblr_pre.js"></script>

</head>
<body>
<div id="wrap">
	<div id="main">
		<div id="header">
			<h1>bio-samtools Basic Tutorial</h1>
		</div>
		<div>
			<h2>Introduction</h2>
			<p>bio-samtools is a Ruby binding to the popular <a href="http://samtools.sourceforge.net/">SAMtools</a> library, and provides access to individual read alignments as well as BAM files, reference sequence and pileup information. </p> 
		</div>
		<div>
			<h2>Installation</h2>
			<p>Installation of bio-samtools is very straightforward, and is accomplished with the Ruby <emph>gems</emph> command. All you need is an internet connection.</p>
			<h3>Prerequisites</h3>
			<p>bio-samtools relies on the following other rubygems:</p>
				<ul> 
					<li><a href="http://rubygems.org/gems/ffi">FFI</a></li>
					<li><a href="http://rubygems.org/gems/bio">bio >= 1.4.1</a> </li>
				</ul>
				Once these are installed, bio-samtools can be installed with
				<pre><code class="ruby">sudo gem install bio-samtools
				 </code></pre>
				It should then be easy to test whether installation went well. Start interactive Ruby (IRB) in the terminal, and type <code>require 'bio-samtools'</code> if the terminal returns <code>true</code> then all is well.
				<pre><code class="ruby">$ irb
>> require 'bio-samtools'
=> true</code></pre>
			<h2>Working with BAM files</h2><br />
			<h3>Creating a new SAM object</h3>
			<p>
				A SAM object represents the alignments in the BAM file, and is very straightforward to create, you will need a sorted BAM file, to access the alignments and a reference sequence in FASTA format to use the reference sequence. The object can be created and opened as follows:</p>
				<pre><code class="ruby">bam = Bio::DB::Sam.new(:bam=>"my_sorted.bam", :fasta=>'ref.fasta')
bam.open</code></pre>
				Opening the file needs only to be done once for multiple operations on it, access to the alignments is random so you don't need to loop over the entries in the file.
			</p>
			<h3>Getting Reference Sequence</h3>
			<p>Retrieving the reference can only be done if the reference has been loaded, which isn't done automatically in order to save memory. Reference need only be loaded once, and is accessed using reference name, start, end in 1-based co-ordinates. A standard Ruby String object is returned.</p>
			<pre><code class="ruby">bam.load_reference 
sequence_fragment = bam.fetch_reference("Chr1", 1, 500)</code></pre>
			<h3>Getting Alignments</h3>
			<p>Alignments can be obtained one at a time by looping over a specified region using the <code>fetch()</code> function.</p>
			<pre><code class="ruby">bam.load_reference 
			bam.fetch("1",3000,4000).each do |alignment|
				#do something with the alignment...
			end</code></pre>			
			<p> A separate method <code>fetch_with_function()</code> allows you to pass a block (or a Proc object) to the function for efficient calculation. This example shows a naive conversion of the alignment object to a GFF3 object, which is stored in an array <code>gff_list</code> </p>
			<pre><code class="ruby">gff_list = []
fetchAlignment = Proc.new do |a|
#what strand is this alignment on...
         a.query_strand ? strand = '+'  : strand = '-'       
         gff_list << Bio::DB::GFF3.new(
           :seqid => "Chr1",
           :start => a.pos - 1,
           :end => a.calend,
           :strand => strand,
           :sequence => a.seq,
           :quality => a.qual,
           :feature => 'read',
           :source => 'BWA',
           :phase => '.',
           :score => '.'
         )   
         0  
       end
       bam.fetch_with_function("Chr1", 3000, 4000, fetchAlignment) #now run the fetch</code></pre>			
			<h4>Alignment Objects</h4>
			<p>The individual alignments represent a single read and are returned as Bio::DB::Alignment objects. These have numerous methods of their own, using <code>require 'pp'</code> will allow you to check the attributes contained in each object. Here is an example alignment object. Remember <code>@</code> represents a Ruby instance variable and can be accessed as any other method. Thus the <code>@is_mapped</code> attribute of an object <code>a</code> is accessed <code>a.is_mapped</code></p>
<pre><code class="ruby">require 'pp'
pp an_alignment_object ##some Bio::DB::Alignment object
#&lt;Bio::DB::Alignment:0x101113f80
@al=#&lt;Bio::DB::SAM::Tools::Bam1T:0x101116a50&gt;,
@calend=4067,
@cigar="76M",
@failed_quality=false,
@first_in_pair=false,
@flag=163,
@is_duplicate=false,
@is_mapped=true,
@is_paired=true,
@isize=180,
@mapq=60,
@mate_strand=false,
@mate_unmapped=false,
@mpos=4096,
@mrnm="=",
@pos=3992,
@primary=true,
@qlen=76,
@qname="HWI-EAS396_0001:7:115:17904:15958#0",
@qual="IIIIIIIIIIIIHHIHGIHIDGGGG...",
@query_strand=true,
@query_unmapped=false,
@rname="1",
@second_in_pair=true,
@seq="ACAGTCCAGTCAAAGTACAAATCGAG...",
@tags=
	{"MD"=>#&lt;Bio::DB::Tag:0x101114ed0 @tag="MD", @type="Z", @value="76"&gt;,
	 "XO"=>#&lt;Bio::DB::Tag:0x1011155d8 @tag="XO", @type="i", @value="0"&gt;,
	 "AM"=>#&lt;Bio::DB::Tag:0x101116280 @tag="AM", @type="i", @value="37"&gt;,
	 "X0"=>#&lt;Bio::DB::Tag:0x101115fb0 @tag="X0", @type="i", @value="1"&gt;,
	 "X1"=>#&lt;Bio::DB::Tag:0x101115c68 @tag="X1", @type="i", @value="0"&gt;,
	 "XG"=>#&lt;Bio::DB::Tag:0x101115240 @tag="XG", @type="i", @value="0"&gt;,
	 "SM"=>#&lt;Bio::DB::Tag:0x1011162f8 @tag="SM", @type="i", @value="37"&gt;,
	 "XT"=>#&lt;Bio::DB::Tag:0x1011162a8 @tag="XT", @type="A", @value="U"&gt;,
	 "NM"=>#&lt;Bio::DB::Tag:0x101116348 @tag="NM", @type="i", @value="0"&gt;,
	 "XM"=>#&lt;Bio::DB::Tag:0x101115948 @tag="XM", @type="i", @value="0"&gt;}&gt;</code></pre>
	<h2>Getting Coverage Information</h2><br />
	<h3>Per Base Coverage</h3>
	<p>It is easy to get the total depth of reads at a given position, the <code>chromosome_coverage</code> function is used. This differs from the previous functions in that a start position and length (rather than end position) are passed to the function. An array of coverages is returned, the first position in the array gives the depth of coverage at the given start position in the genome, the last position in the array gives the depth of coverage at the given start position plus the length given</p>
<pre><code class="ruby">coverages = bam.chromosome_coverage("Chr1", 3000, 1000)  #=> [16,16,25,25...]</code></pre>
<h3>Average Coverage In A Region</h3>
<p>Similarly, average (arithmetic mean) of coverage can be retrieved, also with start and length parameters</p>
<pre><code class="ruby">coverages = bam.average_coverage("Chr1", 3000, 1000)  #=> 20.287</code></pre>

<h3>Getting Pileup Information</h3>
<p>Pileup format represents the coverage of reads over a single base in the reference. Getting a Pileup over a region is very easy. Note that this is done with <code>mpileup</code> and NOT the now deprecated SAMTools <code>pileup</code> function. Calling the <code>mpileup</code> method creates an iterator that yields a Pileup object for each base.</p> 
<pre><code class="ruby">bam.mpileup do |pileup|
	puts pileup.consensus #gives the consensus base from the reads for that postion
end 
</code></pre>
<h4>Pileup options</h4>
<p>The <code>mpileup</code> function takes a range of parameters to allow SAMTools level filtering of reads and alignments. They are specified as key =&gt; value pairs eg</p>
<pre><code class="ruby">bam.mpileup(:r => "Chr1:1000-2000", :Q => 50) do |pileup|
	##only pileups on Chr1 between positions 1000-2000 are considered, 
	##bases with Quality Score < 50 are excluded
	...
end 
</code></pre>	
<p>Not all the options SAMTools allows you to pass to mpileup are supported, those that cause mpileup to return BCF/VCF are ignored. Specifically these are g,u,e,h,I,L,o,p. The table below lists the SAMTools flags supported and the symbols you can use to call them in the mpileup command.
	<table>
		<tr><th>SAMTools option</th><th>description</th><th>short symbol</th><th>long symbol</th><th>default</th><th>example</th></tr>
		<tr><td><code>r</code></td><td>limit retrieval to a region</td><td><code>:r</code></td><td><code>:region</code></td><td>all positions</td><td><code>:r => "Chr1:1000-2000"</code></tr>
		<tr><td><code>6</code></td><td>assume Illumina scaled quality scores</td><td><code>:six</code></td><td><code>:illumina_quals</code></td><td>false</td><td><code>:six => true</code></tr>
		<tr><td><code>A</code></td><td>count anomalous read pairs scores</td><td><code>:A</code></td><td><code>:count_anomalous</code></td><td>false</td><td><code>:A => true</code></tr>
<tr><td><code>B</code></td><td>disable BAQ computation</td><td><code>:B</code></td><td><code>:no_baq</code></td><td>false</td><td><code>:no_baq => true</code></tr>
<tr><td><code>C</code></td><td>parameter for adjusting mapQ</td><td><code>:C</code></td><td><code>:adjust_mapq</code></td><td>0</td><td><code>:C => 25</code></tr>
<tr><td><code>d</code></td><td>max per-BAM depth to avoid excessive memory usage</td><td><code>:d</code></td><td><code>:max_per_bam_depth</code></td><td>250</td><td><code>:d => 123</code></tr>
<tr><td><code>E</code></td><td>extended BAQ for higher sensitivity but lower specificity</td><td><code>:E</code></td><td><code>:extended_baq</code></td><td>false</td><td><code>:E => true</code></tr>
<tr><td><code>G</code></td><td>exclude read groups listed in FILE</td><td><code>:G</code></td><td><code>:exclude_reads_file</code></td><td>false</td><td><code>:G => 'my_file.txt'</code></tr>
<tr><td><code>l</code></td><td>list of positions (chr pos) or regions (BED)</td><td><code>:l</code></td><td><code>:list_of_positions</code></td><td>false</td><td><code>:l => 'my_posns.bed'</code></tr>
<tr><td><code>M</code></td><td>cap mapping quality at value</td><td><code>:M</code></td><td><code>:mapping_quality_cap</code></td><td>60</td><td><code>:M => 40 </code></tr>
<tr><td><code>R</code></td><td>ignore RG tags</td><td><code>:R</code></td><td><code>:ignore_rg</code></td><td>false</td><td><code>:R => true </code></tr>
<tr><td><code>q</code></td><td>skip alignments with mapping quality smaller than value</td><td><code>:q</code></td><td><code>:min_mapping_quality</code></td><td>0</td><td><code>:q => 30 </code></tr>
<tr><td><code>Q</code></td><td>skip bases with base quality smaller than value</td><td><code>:Q</code></td><td><code>:imin_base_quality</code></td><td>13</td><td><code>:Q => 30 </code></tr>
	</table>
		</div>
	</div>
</div>

</body>
</html>
